File: IOS/ALAFIA/Models/CommunityHealth.swift
import Foundation

// MARK: - Community Health Alert

struct CommunityHealthAlert: Codable, Identifiable {
    let id: Int
    let title: String
    let summary: String?
    let description: String?
    let category: String
    let severity: String
    let source: String
    let sourceUrl: String?
    let sourceReferenceId: String?
    let affectedCountries: [String]?
    let affectedRegions: [String]?
    let isGlobal: Bool
    let issuedDate: String?
    let effectiveDate: String?
    let expiryDate: String?
    // Recall fields
    let productName: String?
    let productType: String?
    let manufacturer: String?
    let lotNumbers: [String]?
    let reasonForRecall: String?
    let recallClass: String?
    // Outbreak fields
    let diseaseName: String?
    let pathogen: String?
    let confirmedCases: Int?
    let deaths: Int?
    // Advisory fields
    let recommendedActions: [String]?
    let targetPopulation: [String]?
    // Status
    let isActive: Bool
    let isResolved: Bool
    let resolutionDate: String?
    let resolutionNotes: String?
    let tags: [String]?
    let createdAt: String?
    let updatedAt: String?

    enum CodingKeys: String, CodingKey {
        case id, title, summary, description, category, severity, source, tags
        case sourceUrl = "source_url"
        case sourceReferenceId = "source_reference_id"
        case affectedCountries = "affected_countries"
        case affectedRegions = "affected_regions"
        case isGlobal = "is_global"
        case issuedDate = "issued_date"
        case effectiveDate = "effective_date"
        case expiryDate = "expiry_date"
        case productName = "product_name"
        case productType = "product_type"
        case manufacturer
        case lotNumbers = "lot_numbers"
        case reasonForRecall = "reason_for_recall"
        case recallClass = "recall_class"
        case diseaseName = "disease_name"
        case pathogen
        case confirmedCases = "confirmed_cases"
        case deaths
        case recommendedActions = "recommended_actions"
        case targetPopulation = "target_population"
        case isActive = "is_active"
        case isResolved = "is_resolved"
        case resolutionDate = "resolution_date"
        case resolutionNotes = "resolution_notes"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }

    var categoryIcon: String {
        switch category {
        case "pandemic": return "🦠"
        case "epidemic": return "🏥"
        case "outbreak": return "⚠️"
        case "recall_drug": return "💊"
        case "recall_food": return "🍽️"
        case "recall_device": return "🩺"
        case "advisory": return "📢"
        case "poison": return "☠️"
        case "environmental": return "🌍"
        case "natural_disaster": return "🌪️"
        case "vaccination": return "💉"
        case "guideline_update": return "📋"
        case "travel_health": return "✈️"
        case "water_safety": return "💧"
        case "air_quality": return "🌫️"
        default: return "📌"
        }
    }

    var severityColor: String {
        switch severity {
        case "critical": return "red"
        case "high": return "orange"
        case "moderate": return "yellow"
        case "low": return "green"
        case "info": return "blue"
        default: return "gray"
        }
    }
}

struct AlertCreate: Encodable {
    let title: String
    let summary: String
    let category: String
    let severity: String
    let source: String

    enum CodingKeys: String, CodingKey {
        case title, summary, category, severity, source
    }
}

// MARK: - Community Health Report

struct CommunityHealthReport: Codable, Identifiable {
    let id: Int
    let userId: Int
    let reportType: String
    let title: String
    let description: String?
    let latitude: Double?
    let longitude: Double?
    let locationName: String?
    let country: String?
    let region: String?
    let symptomsReported: [String]?
    let numberAffected: Int?
    let onsetDate: String?
    let productInvolved: String?
    let severity: String?
    let isVerified: Bool
    let isActive: Bool
    let createdAt: String?

    enum CodingKeys: String, CodingKey {
        case id, title, description, latitude, longitude, country, region, severity
        case userId = "user_id"
        case reportType = "report_type"
        case locationName = "location_name"
        case symptomsReported = "symptoms_reported"
        case numberAffected = "number_affected"
        case onsetDate = "onset_date"
        case productInvolved = "product_involved"
        case isVerified = "is_verified"
        case isActive = "is_active"
        case createdAt = "created_at"
    }
}

struct ReportCreate: Encodable {
    var reportType: String = "symptom_cluster"
    var title: String = ""
    var description: String = ""
    var locationName: String?
    var country: String?
    var severity: String = "moderate"
    var numberAffected: Int = 1
    var symptomsReported: [String]?

    enum CodingKeys: String, CodingKey {
        case title, description, country, severity
        case reportType = "report_type"
        case locationName = "location_name"
        case numberAffected = "number_affected"
        case symptomsReported = "symptoms_reported"
    }
}

// MARK: - Health Guideline

struct HealthGuideline: Codable, Identifiable {
    let id: Int
    let title: String
    let summary: String?
    let fullText: String?
    let category: String
    let source: String
    let sourceUrl: String?
    let version: String?
    let effectiveDate: String?
    let supersededDate: String?
    let isCurrent: Bool
    let targetPopulation: [String]?
    let applicableCountries: [String]?
    let applicableConditions: [String]?
    let recommendations: [String]?
    let tags: [String]?
    let createdAt: String?

    enum CodingKeys: String, CodingKey {
        case id, title, summary, category, source, version, tags, recommendations
        case fullText = "full_text"
        case sourceUrl = "source_url"
        case effectiveDate = "effective_date"
        case supersededDate = "superseded_date"
        case isCurrent = "is_current"
        case targetPopulation = "target_population"
        case applicableCountries = "applicable_countries"
        case applicableConditions = "applicable_conditions"
        case createdAt = "created_at"
    }
}

// MARK: - Dashboard Stats

struct CommunityDashboardStats: Codable {
    let totalActiveAlerts: Int
    let criticalAlerts: Int
    let activeRecalls: Int
    let activeOutbreaks: Int
    let activeAdvisories: Int
    let activePoisonAlerts: Int
    let totalGuidelines: Int
    let userReports30d: Int
    let unreadAlerts: Int
    let bookmarkedAlerts: Int
    let latestOutbreaks: [CommunityHealthAlert]
    let latestRecalls: [CommunityHealthAlert]
    let latestAdvisories: [CommunityHealthAlert]

    enum CodingKeys: String, CodingKey {
        case totalActiveAlerts = "total_active_alerts"
        case criticalAlerts = "critical_alerts"
        case activeRecalls = "active_recalls"
        case activeOutbreaks = "active_outbreaks"
        case activeAdvisories = "active_advisories"
        case activePoisonAlerts = "active_poison_alerts"
        case totalGuidelines = "total_guidelines"
        case userReports30d = "user_reports_30d"
        case unreadAlerts = "unread_alerts"
        case bookmarkedAlerts = "bookmarked_alerts"
        case latestOutbreaks = "latest_outbreaks"
        case latestRecalls = "latest_recalls"
        case latestAdvisories = "latest_advisories"
    }
}

// MARK: - Reference Data

struct AlertCategoryInfo: Codable, Identifiable {
    let id: String
    let name: String
    let description: String
    let icon: String
}

struct AlertSourceInfo: Codable, Identifiable {
    let id: String
    let name: String
    let fullName: String
    let url: String?
    let country: String?

    enum CodingKeys: String, CodingKey {
        case id, name, url, country
        case fullName = "full_name"
    }
}

// MARK: - Subscription

struct AlertSubscription: Codable {
    let id: Int?
    let userId: Int?
    let categories: [String]?
    let sources: [String]?
    let countries: [String]?
    let severities: [String]?
    let pushEnabled: Bool
    let emailEnabled: Bool
    let smsEnabled: Bool

    enum CodingKeys: String, CodingKey {
        case id, categories, sources, countries, severities
        case userId = "user_id"
        case pushEnabled = "push_enabled"
        case emailEnabled = "email_enabled"
        case smsEnabled = "sms_enabled"
    }
}

struct SubscriptionCreate: Encodable {
    var categories: [String] = []
    var sources: [String] = []
    var countries: [String] = []
    var severities: [String] = ["critical", "high"]
    var pushEnabled: Bool = true
    var emailEnabled: Bool = false
    var smsEnabled: Bool = false

    enum CodingKeys: String, CodingKey {
        case categories, sources, countries, severities
        case pushEnabled = "push_enabled"
        case emailEnabled = "email_enabled"
        case smsEnabled = "sms_enabled"
    }
}

File: WEB/backend/app/schemas/community_health.py
"""Community Health schemas — alerts, recalls, reports, guidelines, subscriptions."""

from pydantic import BaseModel
from typing import Optional
from datetime import date, datetime


# ─── Alert / Recall / Advisory ────────────────────────────────────────

class AlertCreate(BaseModel):
    title: str
    summary: str
    description: Optional[str] = None
    category: str  # pandemic, recall_drug, recall_food, advisory, poison, etc.
    severity: str = "moderate"  # info, low, moderate, high, critical
    source: str  # who, cdc, fda, ema, health_canada, poison_control, etc.
    source_url: Optional[str] = None
    source_reference_id: Optional[str] = None

    affected_countries: Optional[list[str]] = None
    affected_regions: Optional[list[str]] = None
    is_global: bool = False

    issued_date: date
    effective_date: Optional[date] = None
    expiry_date: Optional[date] = None

    # Recall fields
    product_name: Optional[str] = None
    product_type: Optional[str] = None
    manufacturer: Optional[str] = None
    lot_numbers: Optional[list[str]] = None
    reason_for_recall: Optional[str] = None
    recall_class: Optional[str] = None

    # Outbreak fields
    disease_name: Optional[str] = None
    pathogen: Optional[str] = None
    confirmed_cases: Optional[int] = None
    deaths: Optional[int] = None

    # Advisory fields
    recommended_actions: Optional[list[str]] = None
    target_population: Optional[list[str]] = None

    tags: Optional[list[str]] = None


class AlertUpdate(BaseModel):
    title: Optional[str] = None
    summary: Optional[str] = None
    severity: Optional[str] = None
    confirmed_cases: Optional[int] = None
    deaths: Optional[int] = None
    is_active: Optional[bool] = None
    is_resolved: Optional[bool] = None
    resolution_date: Optional[date] = None
    resolution_notes: Optional[str] = None


class AlertResponse(BaseModel):
    id: int
    title: str
    summary: str
    description: Optional[str] = None
    category: str
    severity: str
    source: str
    source_url: Optional[str] = None
    source_reference_id: Optional[str] = None

    affected_countries: Optional[list[str]] = None
    affected_regions: Optional[list[str]] = None
    is_global: bool

    issued_date: date
    effective_date: Optional[date] = None
    expiry_date: Optional[date] = None

    product_name: Optional[str] = None
    product_type: Optional[str] = None
    manufacturer: Optional[str] = None
    lot_numbers: Optional[list[str]] = None
    reason_for_recall: Optional[str] = None
    recall_class: Optional[str] = None

    disease_name: Optional[str] = None
    pathogen: Optional[str] = None
    confirmed_cases: Optional[int] = None
    deaths: Optional[int] = None

    recommended_actions: Optional[list[str]] = None
    target_population: Optional[list[str]] = None

    is_active: bool
    is_resolved: bool
    resolution_date: Optional[date] = None
    resolution_notes: Optional[str] = None

    tags: Optional[list[str]] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


# ─── Community Health Report ──────────────────────────────────────────

class ReportCreate(BaseModel):
    report_type: str  # symptom_cluster, foodborne_illness, poisoning, etc.
    title: str
    description: str
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    location_name: Optional[str] = None
    country: Optional[str] = None
    region: Optional[str] = None
    symptoms_reported: Optional[list[str]] = None
    number_affected: Optional[int] = None
    onset_date: Optional[date] = None
    product_involved: Optional[str] = None
    severity: str = "moderate"


class ReportResponse(BaseModel):
    id: int
    user_id: int
    report_type: str
    title: str
    description: str
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    location_name: Optional[str] = None
    country: Optional[str] = None
    region: Optional[str] = None
    symptoms_reported: Optional[list[str]] = None
    number_affected: Optional[int] = None
    onset_date: Optional[date] = None
    product_involved: Optional[str] = None
    severity: str
    is_verified: bool
    is_active: bool
    created_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


# ─── Health Guidelines ────────────────────────────────────────────────

class GuidelineCreate(BaseModel):
    title: str
    summary: str
    full_text: Optional[str] = None
    category: str  # infection_prevention, vaccination, chronic_disease, etc.
    source: str
    source_url: Optional[str] = None
    version: Optional[str] = None
    effective_date: Optional[date] = None
    target_population: Optional[list[str]] = None
    applicable_countries: Optional[list[str]] = None
    applicable_conditions: Optional[list[str]] = None
    recommendations: Optional[list[str]] = None
    tags: Optional[list[str]] = None


class GuidelineResponse(BaseModel):
    id: int
    title: str
    summary: str
    full_text: Optional[str] = None
    category: str
    source: str
    source_url: Optional[str] = None
    version: Optional[str] = None
    effective_date: Optional[date] = None
    superseded_date: Optional[date] = None
    is_current: bool
    target_population: Optional[list[str]] = None
    applicable_countries: Optional[list[str]] = None
    applicable_conditions: Optional[list[str]] = None
    recommendations: Optional[list[str]] = None
    tags: Optional[list[str]] = None
    created_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


# ─── User Subscription ───────────────────────────────────────────────

class SubscriptionCreate(BaseModel):
    categories: Optional[list[str]] = None
    sources: Optional[list[str]] = None
    countries: Optional[list[str]] = None
    severities: Optional[list[str]] = None
    push_enabled: bool = True
    email_enabled: bool = False
    sms_enabled: bool = False


class SubscriptionResponse(BaseModel):
    id: int
    user_id: int
    categories: Optional[list[str]] = None
    sources: Optional[list[str]] = None
    countries: Optional[list[str]] = None
    severities: Optional[list[str]] = None
    push_enabled: bool
    email_enabled: bool
    sms_enabled: bool
    created_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


# ─── Dashboard Stats ─────────────────────────────────────────────────

class CommunityDashboardStats(BaseModel):
    """Aggregate stats for the community health dashboard."""
    total_active_alerts: int = 0
    critical_alerts: int = 0
    active_recalls: int = 0
    active_outbreaks: int = 0
    active_advisories: int = 0
    active_poison_alerts: int = 0
    total_guidelines: int = 0
    user_reports_30d: int = 0
    unread_alerts: int = 0
    bookmarked_alerts: int = 0

    # Latest outbreak summary
    latest_outbreaks: list[AlertResponse] = []
    # Latest recalls
    latest_recalls: list[AlertResponse] = []
    # Latest advisories
    latest_advisories: list[AlertResponse] = []

    model_config = {"from_attributes": True}


# ─── Reference Data ──────────────────────────────────────────────────

class AlertCategoryInfo(BaseModel):
    id: str
    name: str
    description: str
    icon: str


class AlertSourceInfo(BaseModel):
    id: str
    name: str
    full_name: str
    url: str
    country: Optional[str] = None
